File: _java-archive/acquisition-service/src/main/java/dev/kritchalach/acquisition/domain/WillingContract.java
package dev.kritchalach.acquisition.domain;

import jakarta.persistence.*;
import java.math.BigDecimal;
import java.time.Instant;
import java.util.UUID;

/**
 * Entity: WillingContract
 * ร่างสัญญาจะซื้อจะขาย (willing contract) ที่ Legal ทำขึ้นเพื่อซื้อ property จาก seller
 * อยู่ภายใน Acquisition Aggregate — access ต้องผ่าน Acquisition Root เท่านั้น
 */
@Entity
@Table(name = "willing_contracts")
public class WillingContract {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID willingContractId;

    private String fileUrl;

    private BigDecimal agreedPrice;

    private UUID templateId;

    @Column(nullable = false)
    private Instant draftedAt;

    protected WillingContract() {}

    public WillingContract(UUID templateId, String fileUrl, BigDecimal agreedPrice) {
        this.templateId = templateId;
        this.fileUrl = fileUrl;
        this.agreedPrice = agreedPrice;
        this.draftedAt = Instant.now();
    }

    public UUID getWillingContractId() { return willingContractId; }
    public String getFileUrl() { return fileUrl; }
    public BigDecimal getAgreedPrice() { return agreedPrice; }
    public UUID getTemplateId() { return templateId; }
    public Instant getDraftedAt() { return draftedAt; }
}

File: acquisition-service/src/domain/WillingContract.ts
import { v4 as uuidv4 } from 'uuid';

/**
 * Entity: WillingContract
 * ร่างสัญญาจะซื้อจะขาย ที่ Legal ทำขึ้นเพื่อซื้อ property จาก seller
 * อยู่ภายใน Acquisition Aggregate
 */
export class WillingContract {
  willingContractId: string;
  fileUrl: string;
  agreedPrice: number;
  templateId: string;
  draftedAt: string;

  private constructor(
    willingContractId: string,
    templateId: string,
    fileUrl: string,
    agreedPrice: number,
    draftedAt: string,
  ) {
    this.willingContractId = willingContractId;
    this.templateId = templateId;
    this.fileUrl = fileUrl;
    this.agreedPrice = agreedPrice;
    this.draftedAt = draftedAt;
  }

  static create(templateId: string, fileUrl: string, agreedPrice: number): WillingContract {
    return new WillingContract(uuidv4(), templateId, fileUrl, agreedPrice, new Date().toISOString());
  }

  static rehydrate(
    willingContractId: string,
    templateId: string,
    fileUrl: string,
    agreedPrice: number,
    draftedAt: string,
  ): WillingContract {
    return new WillingContract(willingContractId, templateId, fileUrl, agreedPrice, draftedAt);
  }
}
